# Task 4.8 

Index:

1.- Importing Data Frames and loading libraries

2.- Comparing the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe

3.- Creating a loyalty flag for existing customers using the transform() and loc() functions.

4.- For: marketing team - difference between the spending habits of the three types of customers

5.- Targeting different types of spenders for marketing campaigns.

6.- Exporting Dataframe


## 1.- Importing Data Frames and loading libraries

In [7]:
# Importing Data: 

import pandas as pd
import numpy as np
import os

path = r'C:\Users\cmfag\OneDrive\Desktop\Data Analyst Specialist\Achievement 4\4.3_orders_products\02 Data\Prepared Data'

df_ords_prods_merge = pd.read_pickle(os.path.join(path, 'df_ords_prods_merge_12_11.pkl'))


In [4]:
df_ords_prods_merge.shape

(32435059, 21)

## 2.- Comparing the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe

Analyze the result. How do the results for the entire dataframe differ from those of the subset? 




In [29]:
#Creating Subset

df_ords_prods_merge_ss = df_ords_prods_merge[:1000000]

df_ords_prods_merge_ss.shape

(1000000, 21)

In [27]:
df_ords_prods_merge_ss.groupby(['department_id']).agg({'order_number':['mean']})

order_number
                      mean
department_id             
1.0               6.567568
2.0               4.000000
3.0               4.235294
4.0               5.675958
5.0               3.000000
6.0               5.875000
7.0               6.700935
8.0               3.857143
9.0               6.000000
11.0              3.500000
12.0              5.360000
13.0              5.742857
14.0              4.555556
15.0              4.960000
16.0              5.955224
17.0              3.133333
18.0              1.333333
19.0              6.138889
20.0              4.815385
21.0              7.000000

In [23]:
# Applying grouoby and agg in the dataframe


df_ords_prods_merge.groupby(['department_id']).agg({'order_number':['mean']})

order_number
                      mean
department_id             
1.0              15.457838
2.0              17.277920
3.0              17.170395
4.0              17.811403
5.0              15.215751
6.0              16.439806
7.0              17.225802
8.0              15.340650
9.0              15.895474
10.0             20.197148
11.0             16.170638
12.0             15.887671
13.0             16.583536
14.0             16.773669
15.0             16.165037
16.0             17.665606
17.0             15.694469
18.0             19.310397
19.0             17.177343
20.0             16.473447
21.0             22.902379

### Result: in the complete dataframe as well as the subset there are 21 depatment_id's and the mean order numbers is lower in the subset compared to the entire data frame.

## 3.- Creating a loyalty flag for existing customers using the transform() and loc() functions.

In [36]:
df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                     NaN       14084                  2          0  ...   
2                     NaN       12427                  3          0  ...   
3                     NaN       26088                  4          0  ...   
4                     NaN       26405                  5          0  ...   

                              product_name aisle_id  department_id  prices  \
0                                     Soda     77.0            7.0     9.0   
1  Organic Unsweetened Vanilla Almond Milk     91.0           16.0    12.5   
2                      Original Beef Jerky     23.0           19.0     4.4   
3               Aged White Cheddar Popcorn     23.0           19.0     4.7   
4         XL Pick-A-Size Paper Towel Rolls     54.0           17.0     1.0   

   merge_check       day_label     busiest_day busiest_period_of_day  \
0         both  Regularly busy  Regularly busy         Fewest orders   
1         both  Regularly busy  Regularly busy         Fewest orders   
2         both  Regularly busy  Regularly busy         Fewest orders   
3         both  Regularly busy  Regularly busy         Fewest orders   
4         both  Regularly busy  Regularly busy         Fewest orders   

  max_order  loyalty_flag  
0        10  New customer  
1        10  New customer  
2        10  New customer  
3        10  New customer  
4        10  New customer  

[5 rows x 21 columns]

In [38]:
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15891507
Loyal customer      10294027
New customer         6249525
Name: count, dtype: int64

In [40]:
df_ords_prods_merge['max_order'].value_counts(dropna = False)

max_order
99    1172313
8      812631
6      812072
9      810842
7      804745
       ...   
93      50440
90      48399
97      44969
98      44604
96      40469
Name: count, Length: 97, dtype: int64

Comments: i created the flag alreay during the exercise, since the flag is included in the dataset as a new culmn name:  loyalty flag, i will just
write the steps here: 

step 1: creating a column with the name max_order - this will help as reference to assign the label to the dataframe

step 2: verify the new column with the name max_order really exist. 

    df_ords_prods_merge.head(5)


step 3_ assign the loyalty flag with the following code_ 

    df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
    df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
    df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'


which means:

    more than 40 orders: loyal customer
    between 10 and 40 orders: regular customer 
    less than 10 orders: nnew customer


step 4: verify the values of the flag by using value counts, result:

    loyalty_flag
    Regular customer    15891507
    Loyal customer      10294027
    New customer         6249525
    Name: count, dtype: int64

step 5: verify the loyalty flag in the dataframe as follows:


In [51]:
df_ords_prods_merge[['user_id', 'loyalty_flag', 'order_number','max_order']].head(10)

user_id  loyalty_flag  order_number  max_order
0        1  New customer             1         10
1        1  New customer             1         10
2        1  New customer             1         10
3        1  New customer             1         10
4        1  New customer             1         10
5        1  New customer             2         10
6        1  New customer             2         10
7        1  New customer             2         10
8        1  New customer             2         10
9        1  New customer             2         10

## 4.- For: marketing team - difference between the spending habits of the three types of customers 
Do prices of products purchased by loyal customers differ from those purchased by regular or new customers?


In [68]:
# Verifiying the spending habits of the three types of customers


df_ords_prods_merge.groupby(['loyalty_flag']).agg({'prices':['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Warning: something is wrong with my data or in the process i did in the previous task, since the max price is by 99999, which can be too high
but for the purposes of this task, I will ignore the error.

Conclusions out of this result: all type of customers have price of products in both the average and max.

In [54]:
#Step 1: identifying average spender

df_ords_prods_merge['average_spend']=df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\cmfag\AppData\Local\Temp\ipykernel_22192\2073435784.py:3: FutureWarning: The provided callable <function mean at 0x0000010D266ED260> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merge['average_spend']=df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [58]:
#Step 2: verify the new column with the name average_spend really exist.

df_ords_prods_merge.head(5)


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                     NaN       14084                  2          0  ...   
2                     NaN       12427                  3          0  ...   
3                     NaN       26088                  4          0  ...   
4                     NaN       26405                  5          0  ...   

   aisle_id department_id  prices  merge_check       day_label  \
0      77.0           7.0     9.0         both  Regularly busy   
1      91.0          16.0    12.5         both  Regularly busy   
2      23.0          19.0     4.4         both  Regularly busy   
3      23.0          19.0     4.7         both  Regularly busy   
4      54.0          17.0     1.0         both  Regularly busy   

      busiest_day busiest_period_of_day max_order  loyalty_flag  average_spend  
0  Regularly busy         Fewest orders        10  New customer       6.367797  
1  Regularly busy         Fewest orders        10  New customer       6.367797  
2  Regularly busy         Fewest orders        10  New customer       6.367797  
3  Regularly busy         Fewest orders        10  New customer       6.367797  
4  Regularly busy         Fewest orders        10  New customer       6.367797  

[5 rows x 22 columns]

In [92]:
# Step 3: assign the spender flag

# Creating flag for low and high spender
df_ords_prods_merge.loc[df_ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

df_ords_prods_merge.loc[df_ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [94]:
# Step 4: Check spender flags
df_ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     31800381
High_spender      634678
Name: count, dtype: int64

In [86]:
pd.options.display.max_rows = 200

In [133]:
df_ords_prods_merge[['user_id', 'order_id','average_spend', 'spender_flag', 'loyalty_flag', 'prices']].head(20)

user_id  order_id  average_spend spender_flag  loyalty_flag  prices
0         1   2539329       6.367797  Low_spender  New customer     9.0
1         1   2539329       6.367797  Low_spender  New customer    12.5
2         1   2539329       6.367797  Low_spender  New customer     4.4
3         1   2539329       6.367797  Low_spender  New customer     4.7
4         1   2539329       6.367797  Low_spender  New customer     1.0
5         1   2398795       6.367797  Low_spender  New customer     9.0
6         1   2398795       6.367797  Low_spender  New customer     3.0
7         1   2398795       6.367797  Low_spender  New customer     4.4
8         1   2398795       6.367797  Low_spender  New customer    10.3
9         1   2398795       6.367797  Low_spender  New customer     4.7
10        1   2398795       6.367797  Low_spender  New customer     4.0
11        1    473747       6.367797  Low_spender  New customer     9.0
12        1    473747       6.367797  Low_spender  New customer     4.4
13        1    473747       6.367797  Low_spender  New customer     3.0
14        1    473747       6.367797  Low_spender  New customer     8.6
15        1    473747       6.367797  Low_spender  New customer     3.2
16        1   2254736       6.367797  Low_spender  New customer     9.0
17        1   2254736       6.367797  Low_spender  New customer     4.4
18        1   2254736       6.367797  Low_spender  New customer     3.0
19        1   2254736       6.367797  Low_spender  New customer     8.6

#### Conclusion: there is not significant difference in the prices of products purchased by loyal customers compared to regular or new customers

## 5.- Targeting different types of spenders for marketing campaigns. 

This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria: If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.” If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [100]:
# Creating a column for frequent customers

df_ords_prods_merge['Customer_frequency'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\cmfag\AppData\Local\Temp\ipykernel_22192\3351417064.py:3: FutureWarning: The provided callable <function median at 0x0000010D268FBEC0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df_ords_prods_merge['Customer_frequency'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [102]:
df_ords_prods_merge.head(5)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                     NaN       14084                  2          0  ...   
2                     NaN       12427                  3          0  ...   
3                     NaN       26088                  4          0  ...   
4                     NaN       26405                  5          0  ...   

   prices merge_check       day_label     busiest_day  busiest_period_of_day  \
0     9.0        both  Regularly busy  Regularly busy          Fewest orders   
1    12.5        both  Regularly busy  Regularly busy          Fewest orders   
2     4.4        both  Regularly busy  Regularly busy          Fewest orders   
3     4.7        both  Regularly busy  Regularly busy          Fewest orders   
4     1.0        both  Regularly busy  Regularly busy          Fewest orders   

  max_order  loyalty_flag average_spend spender_flag  Customer_frequency  
0        10  New customer      6.367797  Low_spender                20.5  
1        10  New customer      6.367797  Low_spender                20.5  
2        10  New customer      6.367797  Low_spender                20.5  
3        10  New customer      6.367797  Low_spender                20.5  
4        10  New customer      6.367797  Low_spender                20.5  

[5 rows x 24 columns]

In [125]:
df_ords_prods_merge.shape

(32435059, 25)

In [119]:
# Creating flag customer_frequency 

df_ords_prods_merge.loc[df_ords_prods_merge['Customer_frequency'] > 20, 'frequency_flag'] = 'Non_frequent_customer'

df_ords_prods_merge.loc[(df_ords_prods_merge['Customer_frequency']>10) & (df_ords_prods_merge['Customer_frequency']<=20),'frequency_flag'] = 'Regular customer'

df_ords_prods_merge.loc[df_ords_prods_merge['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'


In [131]:
df_ords_prods_merge['frequency_flag'].value_counts(dropna=False)

frequency_flag
Frequent Customer        21578273
Regular customer          7217066
Non_frequent_customer     3639720
Name: count, dtype: int64

In [135]:
df_ords_prods_merge[['user_id','days_since_prior_order', 'Customer_frequency', 'frequency_flag']].head(20)

user_id  days_since_prior_order  Customer_frequency         frequency_flag
0         1                     NaN                20.5  Non_frequent_customer
1         1                     NaN                20.5  Non_frequent_customer
2         1                     NaN                20.5  Non_frequent_customer
3         1                     NaN                20.5  Non_frequent_customer
4         1                     NaN                20.5  Non_frequent_customer
5         1                    15.0                20.5  Non_frequent_customer
6         1                    15.0                20.5  Non_frequent_customer
7         1                    15.0                20.5  Non_frequent_customer
8         1                    15.0                20.5  Non_frequent_customer
9         1                    15.0                20.5  Non_frequent_customer
10        1                    15.0                20.5  Non_frequent_customer
11        1                    21.0                20.5  Non_frequent_customer
12        1                    21.0                20.5  Non_frequent_customer
13        1                    21.0                20.5  Non_frequent_customer
14        1                    21.0                20.5  Non_frequent_customer
15        1                    21.0                20.5  Non_frequent_customer
16        1                    29.0                20.5  Non_frequent_customer
17        1                    29.0                20.5  Non_frequent_customer
18        1                    29.0                20.5  Non_frequent_customer
19        1                    29.0                20.5  Non_frequent_customer

## 6.- Exporting Dataframe

In [140]:
df_ords_prods_merge.to_pickle(os.path.join(path,'aggregated_ords_prods_merge.pkl'))